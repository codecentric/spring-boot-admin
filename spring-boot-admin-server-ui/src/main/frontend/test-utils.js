import {render as tlRender} from '@testing-library/vue';
import VueI18n from 'vue-i18n';
import components from '@/components';

export const render = (testComponent, options) => {
  tlRender(testComponent, options,
    vue => {
      vue.use(VueI18n);
      vue.use(components);

      return {
        i18n: new VueI18n({
          missing: (locale, path) => path
        }),
      }
    })
}
