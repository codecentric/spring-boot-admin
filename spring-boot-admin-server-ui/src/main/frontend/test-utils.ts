import NotificationcenterPlugin from '@stekoe/vue-toast-notificationcenter';
import { RenderResult, render as tlRender } from '@testing-library/vue';
import { RouterLinkStub } from '@vue/test-utils';
import { merge } from 'lodash-es';
import { createI18n } from 'vue-i18n';
import { createRouter, createWebHashHistory } from 'vue-router';

import components from './components/index.js';
import SbaModalPlugin from './plugins/modal';

import { createViewRegistry } from '@/composables/ViewRegistry';
import ViewRegistry from '@/viewRegistry';

let terms = {};
const modules: Record<string, any> = import.meta.glob('@/**/i18n.en.json', {
  eager: true,
});
for (const modulesKey in modules) {
  terms = { ...terms, ...modules[modulesKey] };
}
export let router;
createViewRegistry();

export const render = (testComponent, options?): RenderResult => {
  const routes = [{ path: '/', component: testComponent }];
  if (testComponent.install) {
    const viewRegistry = new ViewRegistry();
    testComponent.install({ viewRegistry });
    const routeForComponent = viewRegistry._toRoutes(() => true)[0];

    routes.push({
      ...routeForComponent,
      path: '/' + routeForComponent.path,
    });
  }

  router = createRouter({
    history: createWebHashHistory(),
    routes: routes,
  });

  const renderOptions = merge(
    {
      global: {
        plugins: [
          router,
          createI18n({
            locale: options?.locale || 'en',
            messages: {
              en: terms,
            },
            legacy: false,
            fallbackWarn: false,
            missingWarn: false,
          }),
          NotificationcenterPlugin,
          SbaModalPlugin,
          components,
        ],
        stubs: { RouterLink: RouterLinkStub },
      },
    },
    options,
  );
  return tlRender(testComponent, renderOptions);
};
