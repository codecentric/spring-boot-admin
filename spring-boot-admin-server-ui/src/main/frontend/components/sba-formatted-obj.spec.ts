import { describe } from 'vitest';

import SbaFormattedObj from '@/components/sba-formatted-obj.vue';

import { render } from '@/test-utils';

describe('SbaFormattedObject', () => {
  describe('processes simple text', () => {
    it('autolinks urls', async () => {
      render(SbaFormattedObj, {
        props: {
          value:
            'This is a text with a link https://www.codecentric.de that is automagically created.',
        },
      });

      expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('does render (safe) html', async () => {
      render(SbaFormattedObj, {
        props: {
          value:
            "This is a text with <a href='https://www.codecentric.de'>a link</a> and a bold text <b>bold</b>.",
        },
      });

      expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('does not render unsafe <script> in html', async () => {
      render(SbaFormattedObj, {
        props: {
          value: "<script>(function() { console.log('pwned'); )})();</script>",
        },
      });

      expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('does not render unsafe <img> in html', async () => {
      render(SbaFormattedObj, {
        props: {
          value:
            "<img onload='javascript:alert(1)' src='https://www.codecentric.de' />",
        },
      });

      expect(document.body.innerHTML).toMatchSnapshot();
    });
  });

  describe('processes json', () => {
    it('autolinks urls', async () => {
      render(SbaFormattedObj, {
        props: {
          value: {
            a: {
              nested: {
                object: {
                  key: 'This is a text with a link https://www.codecentric.de that is automagically created.',
                },
              },
            },
          },
        },
      });

      expect(document.body.innerHTML).toMatchSnapshot();
    });

    it('does not render unsafe <script> in html', async () => {
      render(SbaFormattedObj, {
        props: {
          value: {
            a: {
              nested: {
                object: {
                  key: "<script>(function() { console.log('pwned'); )})();</script>",
                },
              },
            },
          },
        },
      });

      expect(document.body.innerHTML).toMatchSnapshot();
    });
  });
});
