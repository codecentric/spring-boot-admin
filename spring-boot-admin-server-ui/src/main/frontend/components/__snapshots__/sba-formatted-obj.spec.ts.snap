// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SbaFormattedObject > processes json > autolinks urls 1`] = `
"<div><pre data-v-4fb4748a="" class="formatted">a:
  nested:
    object:
      key: &gt;-
        This is a text with a link <a href="https://www.codecentric.de" target="_blank">https://www.codecentric.de</a> that is
        automagically created.
</pre></div>"
`;

exports[`SbaFormattedObject > processes json > does not render unsafe <script> in html 1`] = `
"<div><pre data-v-4fb4748a="" class="formatted">a:
  nested:
    object:
      key: 
</pre></div>"
`;

exports[`SbaFormattedObject > processes simple text > autolinks urls 1`] = `"<div><pre data-v-4fb4748a="" class="formatted">This is a text with a link <a href="https://www.codecentric.de" target="_blank">https://www.codecentric.de</a> that is automagically created.</pre></div>"`;

exports[`SbaFormattedObject > processes simple text > does not render unsafe <img> in html 1`] = `"<div><pre data-v-4fb4748a="" class="formatted"></pre></div>"`;

exports[`SbaFormattedObject > processes simple text > does not render unsafe <script> in html 1`] = `"<div><pre data-v-4fb4748a="" class="formatted"></pre></div>"`;

exports[`SbaFormattedObject > processes simple text > does render (safe) html 1`] = `"<div><pre data-v-4fb4748a="" class="formatted">This is a text with <a href="https://www.codecentric.de" target="_blank">a link</a> and a bold text <b>bold</b>.</pre></div>"`;
